_posts/2020-10-13-from-aligned-to-unaligned-checkpoints-part-1.md

```diff
@@ -0,0 +1,117 @@
+
+<span>
+	<center>
+	<img vspace="8" style="width:100%" src="{{site.baseurl}}/img/blog/2020-10-13-from-aligned-to-unaligned-checkpoints-part-1/from-aligned-to-unaligned-checkpoints-part-1-imge-2.png" />
+	</center>
+</span>
+
+Flink’s checkpoint-based approach differs from the approach taken by other stream processing systems, for example keeping state in a distributed database, or writing state changes to a log. The checkpoint-based approach has some nice properties, described below, which make it a great option for Apache Flink.
+
+- **Checkpointing has very simple external dependencies**: An Object Storage or Distributed FileSystem are probably the most available and easiest-to-administer services. Because they are available on all public cloud providers and among the first systems to provide on-premises, Flink becomes a well suited system for a “Cloud Native” stack. Moreover, these storage systems are cheaper by an order of magnitude (GB/month) when compared to distributed databases, key/value stores, or event brokers.
+
+- **Checkpoints are immutable and versioned**: Together with immutable and versioned inputs (input streams are both immutable and version, by nature), they support storing immutable application snapshots that can be used for rollbacks, debugging, testing, or as a cheap alternative to analyze application state outside the production setup.
+
+
+- **Checkpoints decouple the “stream transport” from the persistence mechanism**: “Stream transport” refers to how data is being exchanged between operators, for example during a shuffle. That is key to Flink’s batch <-> streaming unification in one system because it allows Flink to implement a data transport that can take the shape of both a low-latency streaming exchange or a decoupled batch data exchange.
+
+## The Checkpointing Procedure
+
+The fundamental challenge solved by the checkpointing algorithm (details in [this paper](https://pdfs.semanticscholar.org/6fa0/917417d3c213b0e130ae01b7b440b1868dde.pdf)) is drawing a snapshot out of the ever-changing state of a streaming application without suspending the continuous processing of events. Because there are always events in-flight (on the network, in I/O buffers, etc.), up- and downstream operators can be processing events from different times: the sink may write data from 11:04, while the source already ingests events from 11:06. Ideally, all snapshotted data should belong to the same point-in-time, as if the input was paused and we waited until all in-flight data was drained (the pipeline becoming idle) before taking the snapshot.
```

Comment:
```suggestion
The fundamental challenge solved by the checkpointing algorithm (details in [this paper](https://pdfs.semanticscholar.org/6fa0/917417d3c213b0e130ae01b7b440b1868dde.pdf)) is drawing a snapshot out of the ever-changing state of a streaming application without suspending the continuous processing of events. Because there are always events in-flight (on the network, in I/O buffers, etc.), up- and downstream operators can be processing events from different times: the sink may write data from 11:04, while the source already ingests events from 11:06. Ideally, all snapshotted data should belong to the same point-in-time, as if the input was paused and we waited until all in-flight data was drained (i.e. the pipeline becoming idle) before taking the snapshot.
```